### ISSUE TITLE
Add an offline miner to the samplers module

### ISSUE THREAD

**Issue by KevinMusgrave** [KevinMusgrave/pytorch-metric-learning#185]
There should be at least 1 offline miner, even if it's very simple.

**Comment by alpoler** [comment]
I consider adding offline miner. I planned to construct batches composed of anchor_0,positive,negative ...... , anchor_batchsize and put into model and separate them according to order to be employed in triplet margin loss. Is it healthy method ? I check that model should accept only batched data structure rather than list of anchor,pos,negative.

**Comment by alpoler** [comment]
You remove the obligatory of using online miner in this way ? If I can employ this as miner , I guess that it add flexibility to use offline miner Thank you. I am grateful for your help.

**Comment by KevinMusgrave** [comment]
Yes, if you want to train on a pre-determined set of triplets (determined by your offline miner or Sampler) then you have to use the ```EmbeddingsAlreadyPackagedAsTriplets```  to force the loss function to use those triplets.

**Comment by alpoler** [comment]
I have to calculate and concatenate embeddings with help of other defined dataloader to design sampler which takes embeddings as argument but it slows down my implementation. I change train class like i mentioned.

**Comment by KevinMusgrave** [comment]
hmm, I'm not sure what you mean. The sampler's ```__iter__``` should return an iterable of indices. I don't see where the concatenation needs to happen.

**Comment by alpoler** [comment]
Infact, my offline miner select positive or negative sample relatively( random selection , further or nearest to anchor). After that, ı find negative or positive sample which are semihard or hard according to relative sample. Therefore, I have to get embedding in beggining of every epoch to find informative triplet. When  i get embeddings, ı get them from mini batches and concatenate them for sampler along with labels. Sampler takes embedding and labels as input and outputs list of informative triplet in __iter__ method. I'll apply them in your powerful benchmarker to test offline miner. Thanks for your contribution to this libraries. @KevinMusgrave

**Comment by KevinMusgrave** [comment]
Yes it seems reasonable to me that an offline miner would have to compute the current embeddings of the dataset in order to create a new iterable, so I think you're on the right track.

**Comment by fabiozappo** [comment]
Hi @KevinMusgrave and @alpoler 

I am struggling the same problem of get embedding at the begin of every epoch to find informative triplet. Do you have any working code snippet for this?

**Comment by KevinMusgrave** [comment]
@fabiozappo I'm not sure what your setup or exact goal is, but this code might help as an example: https://github.com/KevinMusgrave/pytorch-metric-learning/blob/master/src/pytorch_metric_learning/samplers/tuples_to_weights_sampler.py

The sampler is initialized with the model and the dataset. In ```__iter__``` you compute the embeddings using ```self.model``` and ```self.dataset```, and find the informative triplets. Then you can return flattened indices. For example if you find 5000 triplets, then return 15000 indices. Your training code will have to unflatten the indices to have shape (5000, 3).

Once you've written the sampler, pass it into your dataloader. To ensure that the sampler returns the correct shape, you might want to write your sampler as a batch sampler and pass it into the dataloader as ```batch_sampler```. The difference between a sampler and a batch sampler is that the batch sampler returns batches of indices, so the ```batch_size``` argument of the dataloader becomes irrelevant.
